File: ios/Classes/AdCallbackWrapper.swift
//
//  AdCallbackWrapper.swift
//  clever_ads_solutions
//
//  Created by Владислав Горик on 08.08.2023.
//

import Foundation
import CleverAdsSolutions

class AdCallbackWrapper : CASCallback, CASPaidCallback, CASBannerDelegate {
    
    let flutterCallback: CASFlutterCallback
    let withComplete: Bool
    
    init(flutterCallback: CASFlutterCallback, withComplete: Bool) {
        self.flutterCallback = flutterCallback
        self.withComplete = withComplete
    }
    
    func willShown(ad adStatus: CASImpression) {
        flutterCallback.onShown()
    }
    
    func didShowAdFailed(error: String) {
        flutterCallback.onShowFailed(message: error)
    }
    
    func didClickedAd() {
        flutterCallback.onClicked()
    }
    
    func didCompletedAd() {
        if (withComplete) {
            flutterCallback.onComplete()
        }
    }
    
    func didClosedAd() {
        flutterCallback.onClosed()
    }
    
    func onAdLoaded() {
        flutterCallback.onLoaded()
    }
    
    func onAdFailedToLoad(withError message: String?) {
        flutterCallback.onFailed(error: message)
    }
    
    func bannerAdView(_ adView: CASBannerView, didFailWith casError: CASError) {
        flutterCallback.onFailed(error: casError.message)
    }
    
    func didPayRevenue(for ad: CASImpression) {
        flutterCallback.onImpression(for: ad)
    }
}

File: android/src/main/kotlin/com/cleveradssolutions/plugin/flutter/AdCallbackWrapper.java
package com.cleveradssolutions.plugin.flutter;

import androidx.annotation.NonNull;

import com.cleversolutions.ads.AdError;
import com.cleversolutions.ads.AdPaidCallback;
import com.cleversolutions.ads.AdStatusHandler;
import com.cleversolutions.ads.LoadAdCallback;

public abstract class AdCallbackWrapper implements AdPaidCallback, LoadAdCallback {

    @Override
    public abstract void onShown(@NonNull AdStatusHandler ad);

    @Override
    public abstract void onShowFailed(@NonNull String message);

    @Override
    public abstract void onClicked();

    @Override
    public void onComplete() {
    }

    @Override
    public abstract void onClosed();

    @Override
    public abstract void onAdLoaded();

    @Override
    public void onAdFailedToLoad(@NonNull AdError error) {
        onAdFailed(error.getCode());
    }

    public abstract void onAdFailed(int errorCode);

    @Override
    public abstract void onAdRevenuePaid(@NonNull AdStatusHandler ad);

}